backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java

```diff
@@ -269,98 +194,78 @@ public void addEmail(Email email) {
-                                .collect().in(() -> aggregator, AbstractEmailPayloadAggregator::aggregate).toMulti();
-                    }
+    private Uni<String> renderTitle(Notification item, EmailSubscriptionType emailSubscriptionType, EmailTemplate emailTemplate) {
+        return emailTemplate.getTitle(item.getAction().getEventType(), emailSubscriptionType)
+                .data("action", item.getAction())
+                .createMulti()
+                .collect().with(Collectors.joining())
+                .onFailure()
+                .recoverWithItem(templateEx -> {
+                    log.log(Level.WARNING, templateEx, () -> String.format(
+                            "Unable to render template title for application: [%s], eventType: [%s], subscriptionType: [%s].",
+                            item.getAction().getApplication(),
+                            item.getAction().getEventType(),
+                            emailSubscriptionType
+                    ));
+                    return null;
+                });
+    }
 
-                    if (delete) {
-                        // Nothing to do, delete them right away.
-                        return emailAggregationResources.purgeOldAggregation(aggregationKey, endTime)
-                                .toMulti()
-                                .onItem().transformToMultiAndMerge(i -> Multi.createFrom().empty());
+    private Uni<Tuple3<Email, String, String>> combine(Email email, Uni<String> title, Uni<String> body) {
+        return Uni.combine().all()
+                .unis(
+                        Uni.createFrom().item(email),
+                        title,
+                        body
+                ).asTuple()
+                .onItem().transform(objects -> {
+                    if (objects == null || objects.getItem1() == null || objects.getItem2() == null || objects.getItem3() == null) {
+                        return null;
                     }
 
-                    return Multi.createFrom().empty();
-                })
-                .onItem().transformToMultiAndConcatenate(aggregator -> {
-                    String accountId = aggregationKey.getAccountId();
-                    String bundle = aggregationKey.getBundle();
-                    String application = aggregationKey.getApplication();
-
-                    if (aggregator.getProcessedAggregations() == 0) {
-                        return Multi.createFrom().empty();
-                    }
+                    return objects;
+                });
+    }
 
-                    aggregator.setStartTime(startTime);
-                    aggregator.setEndTimeKey(endTime);
+    @Incoming(AGGREGATION_CHANNEL)
+    public Uni<Void> consumeEmailAggregations(String aggregationsJson) throws JsonProcessingException {
+        List<EmailAggregation> aggregations = objectMapper.readValue(aggregationsJson, new TypeReference<>() {
+        });
```

Comment:
If that line throws a `JsonProcessingException`, the Kafka message won't be acknowledged and there will be another attempt at processing the Kafka message which will likely be invalid again (there's no reason an invalid JSON payload would magically become valid). That sounds like an infinite loop of failed attempts.

If that happens, should we ack the Kafka message and add an ERROR log entry so that we get notified through Sentry that something went wrong and we have to manually fix it? If so, we may need to save the data that was unsuccessfully sent.

Reply:
As far as I understand this here correctly: https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2/acknowledgement/acknowledgement.html

They will get acknowledged with my latest change.

Reply:
The infinite loop is broken indeed. Now there's still the question of the data that may be lost after the scheduled aggregations purge following a failure here.

Reply:
Do you have an idea how to test it?

Reply:
See [this](https://github.com/RedHatInsights/notifications-backend/pull/379#discussion_r674636428) for the test.

I can only think of one case where we could have a payload parsing issue which could cause data loss: it could happen if `backend` and `aggregator` used different versions of the `EmailAggregation` class. As soon as both projects depend on the same version of that class from the `models` module, the problem will be gone.

The data loss isn't such a big deal, I don't think it's worth implementing anything to save that data in case of failure.

Reply:
> I am [one comment](https://github.com/RedHatInsights/notifications-backend/pull/379#discussion_r674651589) away from approving the PR, but be careful with the merge as it will conflict pretty badly with #388 (which I also reviewed). You should talk with @josejulio to determine who's the lucky guy who gets to merge first smile

Yea, I talked to him already :-D luckily he is still sleeping :-D so lets go :-D

Reply:
> I can only think of one case where we could have a payload parsing issue which could cause data loss: it could happen if `backend` and `aggregator` used different versions of the `EmailAggregation` class. As soon as both projects depend on the same version of that class from the `models` module, the problem will be gone.

I just tried to move all the classes to the models project. This doesn't work at the moment. We would need to add reactive things to the aggregator project first.

Reply:
> > I am [one comment](https://github.com/RedHatInsights/notifications-backend/pull/379#discussion_r674651589) away from approving the PR, but be careful with the merge as it will conflict pretty badly with #388 (which I also reviewed). You should talk with @josejulio to determine who's the lucky guy who gets to merge first smile
> 
> Yea, I talked to him already :-D luckily he is still sleeping :-D so lets go :-D

huhu, looking at you @josejulio :-D